spec/forms/support/school_form_spec.rb

```diff
@@ -19,15 +19,79 @@
   describe 'validations' do
     it { is_expected.to be_valid }
 
-    it { is_expected.to validate_presence_of(:location_name).with_message('Enter a name') }
-    it { is_expected.to validate_presence_of(:address1).with_message('Enter address line 1') }
-    it { is_expected.to validate_presence_of(:address3).with_message('Enter a town or city') }
-    it { is_expected.to validate_presence_of(:postcode).with_message('Enter a postcode') }
+    context 'with missing location_name' do
+      it 'is invalid' do
+        params['location_name'] = ''
+        expect(subject).not_to be_valid
+        expect(subject.errors.messages).to eq({ location_name: ['Enter a school name'] })
+      end
+    end
+
+    context 'with missing address1' do
+      it 'is invalid' do
+        params['address1'] = ''
+        expect(subject).not_to be_valid
+        expect(subject.errors.messages).to eq({ address1: ['Enter address line 1'] })
+      end
+    end
+
+    context 'with missing address3' do
+      it 'is invalid' do
+        params['address3'] = ''
+        expect(subject).not_to be_valid
+        expect(subject.errors.messages).to eq({ address3: ['Enter a town or city'] })
+      end
+    end
 
-    it { is_expected.not_to allow_values('tr', 'tr11', 'tr11u').for(:postcode).with_message('Enter a real postcode') }
-    it { is_expected.to allow_value('tr11un').for(:postcode) }
+    context 'with missing postcode' do
+      it 'is invalid' do
+        params['postcode'] = ''
+        expect(subject).not_to be_valid
+        expect(subject.errors.messages).to eq({ postcode: ['Enter a postcode', 'Postcode is not valid (for example, BN1 1AA)'] })
+      end
+    end
 
-    it { is_expected.not_to allow_values('12', '123', '1234', 'qwert').for(:urn).with_message('Site URN must be 5 or 6 numbers') }
+    context 'with invalid postcodes' do
+      it 'is invalid' do
+        params['postcode'] = 'tr1'
+        expect(subject).not_to be_valid
+        expect(subject.errors.messages).to eq({ postcode: ['Postcode is not valid (for example, BN1 1AA)'] })
+
+        params['postcode'] = 'tr11'
+        expect(subject).not_to be_valid
+        expect(subject.errors.messages).to eq({ postcode: ['Postcode is not valid (for example, BN1 1AA)'] })
+
+        params['postcode'] = 'tr11u'
+        expect(subject).not_to be_valid
+        expect(subject.errors.messages).to eq({ postcode: ['Postcode is not valid (for example, BN1 1AA)'] })
+      end
+    end
+
+    context 'with valid postcode' do
+      it 'is valid' do
+        params['postcode'] = 'tr11un'
+        expect(subject).to be_valid
+      end
+    end
+
+    context 'with invalid urns' do
+      it 'is invalid' do
+        params['urn'] = '123'
```

Comment:
Wouldn't using the previous `is_expected.not_to allow_values` and the opposite be a bit more idiomatic for these assertions?

Reply:
When I thought I'd be clever and promote the errors form site with
```
 def validate_site
      skip = []
      return if site.valid?

      sites = @identifier_model.sites.where.not(id: nil)
      skip << :location_name unless sites.exists?(location_name:) || location_name.blank?
      skip << :code unless sites.exists?(code:)

      promote_errors(site.errors, skip)
    end

    def promote_errors(site_errors, skip)
      site_errors.each do |site_error|
        next if skip.include?(site_error.attribute)

        errors.add(site_error.attribute, site_error.message)
      end
    end
```
the matchers stopped working?